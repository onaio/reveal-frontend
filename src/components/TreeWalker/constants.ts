export const ACTIVE = 'Active';
export const FEATURE = 'Feature';
export const FIND_BY_ID = 'location/findByJurisdictionIds';
export const FIND_BY_PROPERTIES = 'location/findByProperties';
export const LOCATION = 'location';
export const COULDNT_LOAD_CHILDREN = 'Could not load children';
export const COULDNT_LOAD_PARENTS = 'Could not load parents';
export const NOT_AVAILABLE = 'N/A';
