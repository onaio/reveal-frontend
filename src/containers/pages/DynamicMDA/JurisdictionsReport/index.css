.generic-report-table .thead .th {
  padding: 10px 10px;
}
