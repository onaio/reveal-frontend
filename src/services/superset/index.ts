import reducerRegistry from '@onaio/redux-reducer-registry';
import superset, { SupersetConnectorConfig } from '@onaio/superset-connector';
import { OPENSRP_OAUTH_STATE, SUPERSET_API_BASE, SUPERSET_API_ENDPOINT } from '../../configs/env';
import { FlexObject } from '../../helpers/utils';
import store from '../../store';
import supersetReducer, {
  authorizeSuperset,
  isAuthorized,
  reducerName as supersetReducerName,
} from '../../store/ducks/superset';
import { getAccessToken, getOauthProviderState } from '../../store/selectors';

// /** register the superset reducer */
reducerRegistry.register(supersetReducerName, supersetReducer);

/** middleware for fetching from Superset */
export const fetchMiddleware = (res: { [key: string]: any }) => {
  return res;
};

/** callback for fetching from Superset */
export const fetchCallback = (parsedResponse: Array<{ [key: string]: any }>) => {
  const sliceData = superset.processData(parsedResponse);
  return sliceData;
};

/** this function completes the authZ process */
export const completeAuthZ = async (result: { [key: string]: any }) => {
  if (result.status === 200) {
    store.dispatch(authorizeSuperset(true));
  } else {
    store.dispatch(authorizeSuperset(false));
  }
};

/** generic async function for fetching from superset */
const supersetFetch = async (
  sliceId: string,
  params: FlexObject | null = null,
  callback: typeof fetchCallback = fetchCallback,
  middleware: typeof fetchMiddleware = fetchMiddleware
) => {
  const config: SupersetConnectorConfig = {
    base: SUPERSET_API_BASE,
    endpoint: SUPERSET_API_ENDPOINT,
    extraPath: sliceId,
    /** the provider should be the name of the provider
     * to enable the use of multiple providers, we store the provider name in the
     * `state` parameter so that we can use it here
     */
    provider: getOauthProviderState(store.getState()) || OPENSRP_OAUTH_STATE,
    token: '',
  };

  const accessToken = getAccessToken(store.getState());
  if (accessToken) {
    config.token = accessToken;
  }

  if (params) {
    config.params = `form_data=${JSON.stringify(params)}`;
  }

  const isSupersetAuthorized = isAuthorized(store.getState());

  if (isSupersetAuthorized === true) {
    return superset.api.doFetch(config, middleware).then(callback);
  } else {
    return superset.authZ(config, (result: { [key: string]: any }) => {
      return completeAuthZ(result)
        .then(() => {
          const isSupersetAuthorizedYet = isAuthorized(store.getState());
          if (isSupersetAuthorizedYet === true) {
            return superset.api.doFetch(config, middleware).then(callback);
          }
        })
        .catch(() => {
          return false;
        });
    });
  }
};

export default supersetFetch;
