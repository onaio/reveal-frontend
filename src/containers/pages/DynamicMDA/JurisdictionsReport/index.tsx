import { hasChildrenFunc } from '@onaio/drill-down-table';
import reducerRegistry from '@onaio/redux-reducer-registry';
import React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { Store } from 'redux';
import GenericTableCell from '../../../../components/TableCells/JurisdictionReport';
import {
  SUPERSET_DYNAMIC_MDA_REPORTING_FOCUS_AREAS_COLUMNS,
  SUPERSET_DYNAMIC_MDA_REPORTING_JURISDICTIONS_COLUMNS,
  SUPERSET_DYNAMIC_MDA_REPORTING_JURISDICTIONS_DATA_SLICES,
  SUPERSET_DYNAMIC_MDA_REPORTING_JURISDICTIONS_FOCUS_AREA_LEVEL,
  SUPERSET_DYNAMIC_MDA_REPORTING_PLANS_SLICE,
} from '../../../../configs/env';
import { MDA_REPORTING_TITLE } from '../../../../configs/lang';
import { REPORT_MDA_PLAN_URL } from '../../../../constants';
import '../../../../helpers/tables.css';
import { RouteParams } from '../../../../helpers/utils';
import supersetFetch from '../../../../services/superset';
import {
  fetchGenericJurisdictions,
  GenericJurisdiction,
  getGenericJurisdictionsArray,
} from '../../../../store/ducks/generic/jurisdictions';
import MDAPlansReducer, {
  genericFetchPlans,
  getPlanByIdSelector,
  reducerName as MDAPlansReducerName,
} from '../../../../store/ducks/generic/plans';
import { GenericPlan } from '../../../../store/ducks/generic/plans';
import {
  GenericJurisdictionProps,
  GenericJurisdictionReport,
} from '../../GenericJurisdictionReport';
import './index.css';

/** register the reducers */
reducerRegistry.register(MDAPlansReducerName, MDAPlansReducer);

/** Renders MDA Jurisdictions reports */
const MDAJurisdictionReport = (
  props: GenericJurisdictionProps & RouteComponentProps<RouteParams>
) => {
  return (
    <div>
      <GenericJurisdictionReport {...props} />
    </div>
  );
};

const defaultProps: GenericJurisdictionProps = {
  LegendIndicatorComp: null,
  baseURL: REPORT_MDA_PLAN_URL,
  cellComponent: GenericTableCell,
  fetchJurisdictions: fetchGenericJurisdictions,
  fetchPlans: genericFetchPlans,
  focusAreaColumn: SUPERSET_DYNAMIC_MDA_REPORTING_FOCUS_AREAS_COLUMNS,
  focusAreaLevel: SUPERSET_DYNAMIC_MDA_REPORTING_JURISDICTIONS_FOCUS_AREA_LEVEL,
  hasChildren: hasChildrenFunc,
  jurisdictionColumn: SUPERSET_DYNAMIC_MDA_REPORTING_JURISDICTIONS_COLUMNS,
  jurisdictions: null,
  pageTitle: MDA_REPORTING_TITLE,
  plan: null,
  reportingPlanSlice: SUPERSET_DYNAMIC_MDA_REPORTING_PLANS_SLICE,
  service: supersetFetch,
  slices: SUPERSET_DYNAMIC_MDA_REPORTING_JURISDICTIONS_DATA_SLICES.split(','),
};

MDAJurisdictionReport.defaultProps = defaultProps;

export { MDAJurisdictionReport };

/** Connect the component to the store */

/** interface to describe props from mapStateToProps */
interface DispatchedStateProps extends RouteComponentProps<RouteParams> {
  jurisdictions: GenericJurisdiction[] | null;
  plan: GenericPlan | null;
}

/** map state to props */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: GenericJurisdictionProps & RouteComponentProps<RouteParams>
): DispatchedStateProps => {
  const planId = ownProps.match.params.planId || null;
  const plan = planId ? getPlanByIdSelector(state, planId) : null;

  let jurisdictions: GenericJurisdiction[] = [];
  defaultProps.slices.forEach(
    slice =>
      (jurisdictions = jurisdictions.concat(getGenericJurisdictionsArray(state, slice, planId)))
  );

  return {
    ...ownProps,
    jurisdictions,
    plan,
  };
};

/** map dispatch to props */
const mapDispatchToProps = {
  fetchJurisdictions: fetchGenericJurisdictions,
  fetchPlans: genericFetchPlans,
};

/** Connected ActiveFI component */
const ConnectedMDAJurisdictionReport = connect(
  mapStateToProps,
  mapDispatchToProps
)(MDAJurisdictionReport);

export default ConnectedMDAJurisdictionReport;
