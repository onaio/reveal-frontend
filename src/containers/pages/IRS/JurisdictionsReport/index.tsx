import DrillDownTable, { hasChildrenFunc } from '@onaio/drill-down-table';
import reducerRegistry from '@onaio/redux-reducer-registry';
import superset, { SupersetFormData } from '@onaio/superset-connector';
import { get } from 'lodash';
import React, { SyntheticEvent, useEffect, useState } from 'react';
import { Helmet } from 'react-helmet';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { Column, RowInfo } from 'react-table';
import 'react-table/react-table.css';
import { Col, Row } from 'reactstrap';
import { Store } from 'redux';
import IRSIndicatorLegend from '../../../../components/formatting/IRSIndicatorLegend';
import IRSTableCell from '../../../../components/IRSTableCell';
import HeaderBreadcrumb from '../../../../components/page/HeaderBreadcrumb/HeaderBreadcrumb';
import Loading from '../../../../components/page/Loading';
import {
  SUPERSET_IRS_REPORTING_FOCUS_AREAS_COLUMNS,
  SUPERSET_IRS_REPORTING_JURISDICTIONS_COLUMNS,
  SUPERSET_IRS_REPORTING_JURISDICTIONS_DATA_SLICES,
  SUPERSET_IRS_REPORTING_JURISDICTIONS_FOCUS_AREA_LEVEL,
  SUPERSET_IRS_REPORTING_PLANS_SLICE,
  SUPERSET_MAX_RECORDS,
} from '../../../../configs/env';
import { HOME, IRS_REPORTING_TITLE } from '../../../../configs/lang';
import { HOME_URL, REPORT_IRS_PLAN_URL } from '../../../../constants';
import { displayError } from '../../../../helpers/errors';
import '../../../../helpers/tables.css';
import { FlexObject, RouteParams } from '../../../../helpers/utils';
import supersetFetch from '../../../../services/superset';
import GenericJurisdictionsReducer, {
  fetchGenericJurisdictions,
  GenericJurisdiction,
  getGenericJurisdictionsArray,
  reducerName as GenericJurisdictionsReducerName,
} from '../../../../store/ducks/generic/jurisdictions';
import IRSPlansReducer, {
  fetchIRSPlans,
  getIRSPlanById,
  IRSPlan,
  reducerName as IRSPlansReducerName,
} from '../../../../store/ducks/generic/plans';
import { getJurisdictionBreadcrumbs } from '../Map/helpers';
import { IRSTableColumns } from './helpers';
import './style.css';

/** register the reducers */
reducerRegistry.register(IRSPlansReducerName, IRSPlansReducer);
reducerRegistry.register(GenericJurisdictionsReducerName, GenericJurisdictionsReducer);

const slices = SUPERSET_IRS_REPORTING_JURISDICTIONS_DATA_SLICES.split(',');

/** IRS Jurisdictions props */
export interface GenericJurisdictionProps {
  fetchJurisdictions: typeof fetchGenericJurisdictions;
  fetchPlans: typeof fetchIRSPlans;
  hasChildren: typeof hasChildrenFunc;
  jurisdictions: GenericJurisdiction[] | null;
  plan: IRSPlan | null;
  service: typeof supersetFetch;
}

/** Renders IRS Jurisdictions reports */
const JurisdictionReport = (props: GenericJurisdictionProps & RouteComponentProps<RouteParams>) => {
  const [jurisdictionId, setJurisdictionId] = useState<string | null>(
    (props.match && props.match.params && props.match.params.jurisdictionId) || null
  );
  const [loading, setLoading] = useState<boolean>(true);

  let planId: string | null = null;
  if (props.match && props.match.params && props.match.params.planId) {
    planId = props.match.params.planId;
  }
  const { fetchJurisdictions, fetchPlans, hasChildren, jurisdictions, plan, service } = props;

  /** async function to load the data */
  async function loadData() {
    try {
      setLoading(!plan || !jurisdictions || jurisdictions.length < 1); // set loading when there is no data
      let fetchPlansParams: SupersetFormData | null = null;
      if (planId) {
        fetchPlansParams = superset.getFormData(1, [
          { comparator: planId, operator: '==', subject: 'plan_id' },
        ]);
      }

      await service(
        SUPERSET_IRS_REPORTING_PLANS_SLICE,
        fetchPlansParams
      ).then((result: IRSPlan[]) => fetchPlans(result));

      slices.forEach(async slice => {
        let fetchJurisdictionsParams: SupersetFormData | null = null;
        if (planId) {
          fetchJurisdictionsParams = superset.getFormData(
            SUPERSET_MAX_RECORDS,
            [{ comparator: planId, operator: '==', subject: 'plan_id' }],
            { jurisdiction_depth: true, jurisdiction_name: true }
          );
        }
        await service(slice, fetchJurisdictionsParams).then((result: GenericJurisdiction[]) =>
          fetchJurisdictions(slice, result)
        );
      });
    } catch (e) {
      // todo - handle error https://github.com/onaio/reveal-frontend/issues/300
    } finally {
      setLoading(false);
    }
  }

  useEffect(() => {
    loadData().catch(err => displayError(err));
  }, []);

  useEffect(() => {
    if (props.match && props.match.params && props.match.params.jurisdictionId) {
      setJurisdictionId(props.match.params.jurisdictionId);
    } else {
      setJurisdictionId(null);
    }
  }, [props.match.params.jurisdictionId]);

  if (loading === true) {
    return <Loading />;
  }

  const data = jurisdictions || [];

  const parentNodes = data.map((el: FlexObject) => el.jurisdiction_parent_id);

  let pageTitle = IRS_REPORTING_TITLE;
  let baseURL = REPORT_IRS_PLAN_URL;
  const basePage = {
    label: pageTitle,
    url: REPORT_IRS_PLAN_URL,
  };
  const breadcrumbProps = {
    currentPage: basePage,
    pages: [
      {
        label: HOME,
        url: HOME_URL,
      },
    ],
  };

  let planPage = basePage;
  if (plan) {
    pageTitle = `${IRS_REPORTING_TITLE}: ${plan.plan_title}`;
    baseURL = `${REPORT_IRS_PLAN_URL}/${plan.plan_id}`;
    planPage = {
      label: plan.plan_title,
      url: baseURL,
    };
    breadcrumbProps.currentPage = planPage;
    breadcrumbProps.pages.push(basePage);
  }

  const theObject = data.filter((el: FlexObject) => el.jurisdiction_id === jurisdictionId);

  let currentJurisdictionName: string | null = null;
  if (theObject && theObject.length > 0) {
    const pages = getJurisdictionBreadcrumbs(theObject[0], baseURL);

    breadcrumbProps.pages.push(planPage);

    const newPages = breadcrumbProps.pages.concat(pages);
    breadcrumbProps.pages = newPages;

    const currentPage = {
      label: theObject[0].jurisdiction_name,
      url: `${baseURL}/${theObject[0].jurisdiction_id}`,
    };
    breadcrumbProps.currentPage = currentPage;

    currentJurisdictionName = theObject[0].jurisdiction_name;
  }

  const currLevelData = data.filter(el => el.jurisdiction_parent_id === jurisdictionId);

  let columnsToUse = get(IRSTableColumns, SUPERSET_IRS_REPORTING_JURISDICTIONS_COLUMNS, null);
  if (currLevelData && currLevelData.length > 0) {
    if (
      currLevelData[0].jurisdiction_depth === +SUPERSET_IRS_REPORTING_JURISDICTIONS_FOCUS_AREA_LEVEL
    ) {
      columnsToUse = get(IRSTableColumns, SUPERSET_IRS_REPORTING_FOCUS_AREAS_COLUMNS, null);
    }
  }

  const tableProps = {
    ...(columnsToUse && { columns: columnsToUse }),
    CellComponent: IRSTableCell,
    data,
    defaultPageSize: data.length,
    extraCellProps: { urlPath: baseURL },
    getTdProps: (_: Partial<Store>, rowInfo: RowInfo | undefined, column: Column | undefined) => {
      return {
        onClick: (__: SyntheticEvent, handleOriginal: () => void) => {
          if (rowInfo && column) {
            if (
              column.id === 'jurisdiction_name' &&
              hasChildren(rowInfo, parentNodes, 'jurisdiction_id')
            ) {
              setJurisdictionId(rowInfo.original.jurisdiction_id);
            }
            if (handleOriginal) {
              handleOriginal();
            }
          }
        },
      };
    },
    identifierField: 'jurisdiction_id',
    linkerField: 'jurisdiction_name',
    minRows: 0,
    parentIdentifierField: 'jurisdiction_parent_id',
    resizable: true,
    rootParentId: jurisdictionId || '',
    shouldUseEffect: false,
    showPagination: false,
    useDrillDownTrProps: false,
  };

  const currentTitle = currentJurisdictionName
    ? `${pageTitle}: ${currentJurisdictionName}`
    : pageTitle;

  return (
    <div key={`${jurisdictionId || '0'}-${data.length}`}>
      <Helmet>
        <title>{currentTitle}</title>
      </Helmet>
      <HeaderBreadcrumb {...breadcrumbProps} />
      <Row>
        <Col>
          <h3 className="mb-3 page-title">{currentTitle}</h3>
          <div className="irs-report-table">
            <DrillDownTable {...tableProps} />
          </div>
          <IRSIndicatorLegend />
        </Col>
      </Row>
    </div>
  );
};

const defaultProps: GenericJurisdictionProps = {
  fetchJurisdictions: fetchGenericJurisdictions,
  fetchPlans: fetchIRSPlans,
  hasChildren: hasChildrenFunc,
  jurisdictions: null,
  plan: null,
  service: supersetFetch,
};

JurisdictionReport.defaultProps = defaultProps;

export { JurisdictionReport };

/** Connect the component to the store */

/** interface to describe props from mapStateToProps */
interface DispatchedStateProps {
  plan: IRSPlan | null;
  jurisdictions: GenericJurisdiction[] | null;
}

/** map state to props */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: GenericJurisdictionProps & RouteComponentProps<RouteParams>
): DispatchedStateProps => {
  const planId = ownProps.match.params.planId || null;
  const plan = planId ? getIRSPlanById(state, planId) : null;
  let jurisdictions: GenericJurisdiction[] = [];

  slices.forEach(
    slice =>
      (jurisdictions = jurisdictions.concat(getGenericJurisdictionsArray(state, slice, planId)))
  );

  return {
    jurisdictions,
    plan,
  };
};

/** map dispatch to props */
const mapDispatchToProps = {
  fetchJurisdictions: fetchGenericJurisdictions,
  fetchPlans: fetchIRSPlans,
};

/** Connected ActiveFI component */
const ConnectedJurisdictionReport = connect(
  mapStateToProps,
  mapDispatchToProps
)(JurisdictionReport);

export default ConnectedJurisdictionReport;
