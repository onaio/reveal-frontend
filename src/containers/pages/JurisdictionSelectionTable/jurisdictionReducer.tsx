/** a custom hook to manage changes to the jurisdiction tree */
import { Failure, Result } from '@onaio/utils/dist/types/types';
import { cloneDeep, uniqWith } from 'lodash';
import React from 'react';
import TreeModel from 'tree-model/types';
import { LoadOpenSRPHierarchy } from '../../../helpers/dataLoading/jurisdictions';
import { failure } from '../../../helpers/dataLoading/utils';
import { OpenSRPService } from '../../../services/opensrp';
import { generateJurisdictionTree, ParsedHierarchySingleNode, RawOpenSRPHierarchy } from './utils';

/** string to use as key when setting value of a node as selected */
const SELECT_KEY = 'selected';

/** given an array of the childrenNodes, return true if all of them are selected
 * @param childrenNodes - an array of nodes
 */
export function areAllChildrenSelected(
  childrenNodes: Array<TreeModel.Node<ParsedHierarchySingleNode>>
) {
  let selected = true;
  childrenNodes.forEach(node => {
    selected = selected && node.model.node.attributes[SELECT_KEY];
  });
  return selected;
}

/** returns if node is selected
 * @param node - The node
 */
export const nodeIsSelected = (node: TreeModel.Node<ParsedHierarchySingleNode>) => {
  return !!node.model.node.attributes[SELECT_KEY];
};

/** checks whether a node is a parent node or a leaf node
 * @param node - the node
 * @return true if node is a parent node, false if node is a leaf node
 */
export const nodeHasChildren = (node: TreeModel.Node<ParsedHierarchySingleNode>) => {
  return node.children && node.children.length > 0;
};

/** constructs, and manages interactions that mutate the tree
 * @param rootJurisdictionId - uses this jurisdiction id to get the opensrp hierarchy
 */
export function useJurisdictionTree(rootJurisdictionId: string) {
  const [tree, setTree] = React.useState<TreeModel.Node<ParsedHierarchySingleNode> | undefined>(
    undefined
  );
  const [currentParentNode, setCurrentParent] = React.useState<
    TreeModel.Node<ParsedHierarchySingleNode> | undefined
  >(undefined);
  const [currentChildren, setCurrentChildren] = React.useState<
    Array<TreeModel.Node<ParsedHierarchySingleNode>>
  >([]);
  const [errors, setErrors] = React.useState<Failure[]>([]);
  const addError = (fail: Failure) => {
    setErrors([...errors, fail]);
  };
  const clearErrors = () => setErrors([]);

  React.useEffect(() => {
    const params = {
      return_structure_count: true,
    };
    LoadOpenSRPHierarchy(rootJurisdictionId, OpenSRPService, params)
      .then((apiResponse: Result<RawOpenSRPHierarchy>) => {
        // create parseable tree using treeModel lib
        if (apiResponse.value) {
          const responseData = apiResponse.value;
          const theTree = generateJurisdictionTree(responseData);
          setTree(theTree);
          setCurrentChildren([theTree]);
        } else {
          addError(apiResponse);
        }
      })
      .catch((err: Error) => {
        addError(failure(err));
      });
  }, []);

  /** make currentChildren reactive depending on the set parent node */
  React.useEffect(() => {
    if (currentParentNode) {
      setCurrentChildren(currentParentNode.children);
    }
    /** initially on mount currentParent node is undefined */
    if (!currentParentNode && tree) {
      setCurrentChildren([tree]);
    }
  }, [currentParentNode]);

  /** Update parent node once the tree is changed */
  React.useEffect(() => {
    if (tree && !currentParentNode) {
      setCurrentChildren([tree]);
    }
    if (currentParentNode) {
      const updatedCurrentParent = tree!.first(nd => nd.model.id === currentParentNode.model.id);
      setCurrentParent(updatedCurrentParent);
    }
  }, [tree]);

  /** helper that sets attributes to a node such that the tree changes causing a re-render
   * @param aTree - the tree, usually the tree state variable
   * @param nodeId - id of the node of interest
   * @param attrAccessor - the key to use when assigning the value
   * @param attrValue - the value to assign to attrAccessor
   * @param cascadeDown - whether to do the value assignment on each of the node's descendants
   * @param cascadeDown - whether to do the value assignment on each of the node's ancestors
   */
  function _setAttrsToNode(
    aTree: TreeModel.Node<ParsedHierarchySingleNode>,
    nodeId: string,
    attrAccessor: string,
    attrValue: string | number | boolean,
    cascadeDown: boolean = false,
    cascadeUp: boolean = false
  ) {
    const treeClone = cloneDeep(aTree);
    const argNode = treeClone.first(nd => nd.model.id === nodeId);
    if (!argNode) {
      return treeClone;
    }
    argNode.model.node.attributes[attrAccessor] = attrValue;
    if (cascadeDown) {
      // argNode is part of this walk.
      argNode.walk(nd => {
        nd.model.node.attributes[attrAccessor] = attrValue;
        // removing this return cause a type error, that's its sole purpose
        return true;
      });
    }
    if (cascadeUp) {
      const parentsPath = argNode.getPath();
      // remove the node from path
      parentsPath.pop();
      parentsPath.forEach(parentNode => {
        parentNode.model.node.attributes[attrAccessor] = attrValue;
      });
    }
    return { treeClone, argNode };
    // do not setTree here yet so that the caller can perform additional mutations
  }

  /** traverse up and select the parent nodes where necessary
   * @param node a node
   */
  function _computeParentNodesSelection(node: TreeModel.Node<ParsedHierarchySingleNode>) {
    const parentsPath = node.getPath();
    parentsPath.pop();
    const reversedParentSPath = parentsPath.reverse();
    // now for each of the parent if all the children are selected then label the parent as selected too
    for (const parentNode of reversedParentSPath) {
      const allChildrenAreSelected = areAllChildrenSelected(parentNode.children);
      if (allChildrenAreSelected) {
        parentNode.model.node.attributes[SELECT_KEY] = allChildrenAreSelected;
      } else {
        break;
      }
    }
  }

  /** marks a node as selected
   * @param nodeId: id of node
   */
  const selectNode = (nodeId: string) => {
    if (!tree) {
      return;
    }
    const { treeClone } = _setAttrsToNode(tree, nodeId, SELECT_KEY, true, true);
    // go up the tree look at the parent's children
    const node = treeClone.first(
      (nd: TreeModel.Node<ParsedHierarchySingleNode>) => nd.model.id === nodeId
    );
    if (!node) {
      // if node was not found do nothing.
      setTree(treeClone);
      return;
    }
    _computeParentNodesSelection(node);
    setTree(treeClone);
  };

  /** will unselect a node and cascade the effect up
   * @param nodeId - the node's id
   */
  const unSelectNode = (nodeId: string) => {
    if (!tree) {
      return;
    }
    const { treeClone } = _setAttrsToNode(tree, nodeId, SELECT_KEY, false, true, true);

    setTree(treeClone);
  };

  /** decides whether to call select or unselect based on the value to be assigned to the selected key
   * @param node - set selected value on this node
   * @param value - the node's selected value to this
   */
  const applySelectToNode = (node: TreeModel.Node<ParsedHierarchySingleNode>, value: boolean) => {
    if (value) {
      selectNode(node.model.id);
    } else {
      unSelectNode(node.model.id);
    }
  };

  /** any callback that will take a node and return whether we should select the node
   * @param callback - callback is given each node in a walk and decides whether that node should be selected by returning true or false
   */
  const autoSelectNodes = (
    callback: (node: TreeModel.Node<ParsedHierarchySingleNode>) => boolean
  ) => {
    const treeClone = cloneDeep(tree);
    if (!treeClone) {
      return;
    }
    const parentNodes: Array<TreeModel.Node<ParsedHierarchySingleNode>> = [];
    treeClone.walk(node => {
      if (callback(node)) {
        node.model.node.attributes[SELECT_KEY] = true;
        parentNodes.push(node.parent);
      }
      return true;
    });

    // remove duplicates
    const parentNodesSet = uniqWith(parentNodes, node => node.model.id);

    // now select parents accordingly
    parentNodesSet.forEach(node => {
      _computeParentNodesSelection(node);
    });
    setTree(treeClone);
  };

  /** returns all selected nodes
   * @param leafNodesOnly - whether to return leaf nodes only
   */
  const getAllSelectedNodes = (leafNodesOnly = false) => {
    const nodesList: Array<TreeModel.Node<ParsedHierarchySingleNode>> = [];
    if (!tree) {
      return [];
    }
    tree.walk(node => {
      const shouldAddNode = nodeIsSelected && !(leafNodesOnly && nodeHasChildren(node));
      if (shouldAddNode) {
        nodesList.push(node);
      }
      return true;
    });
    return nodesList;
  };

  // we need a way to set a current node as selected,
  return {
    applySelectToNode,
    autoSelectNodes,
    clearErrors,
    currentChildren,
    currentParentNode,
    errors,
    getAllSelectedNodes,
    selectNode,
    setCurrentParent,
    tree,
    unSelectNode,
  };
}
