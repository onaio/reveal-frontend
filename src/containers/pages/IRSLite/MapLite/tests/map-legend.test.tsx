import reducerRegistry from '@onaio/redux-reducer-registry';
import superset from '@onaio/superset-connector';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import flushPromises from 'flush-promises';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Router } from 'react-router';
import { IRSReportingMap } from '../';
import { MAP, REPORT_IRS_PLAN_URL } from '../../../../../constants';
import store from '../../../../../store';
import GenericJurisdictionsReducer, {
  fetchGenericJurisdictions,
  getGenericJurisdictionByJurisdictionId,
  reducerName as GenericJurisdictionsReducerName,
} from '../../../../../store/ducks/generic/jurisdictions';
import IRSPlansReducer, {
  GenericPlan,
  reducerName as IRSPlansReducerName,
} from '../../../../../store/ducks/generic/plans';
import genericStructuresReducer, {
  fetchGenericStructures,
  getGenericStructures,
  reducerName as genericStructuresReducerName,
} from '../../../../../store/ducks/generic/structures';
import { plans } from '../../../../../store/ducks/generic/tests/fixtures';
import jurisdictionReducer, {
  fetchJurisdictions,
  getJurisdictionById,
  reducerName as jurisdictionReducerName,
} from '../../../../../store/ducks/jurisdictions';
import * as fixtures from '../../JurisdictionsReport/fixtures';

/* tslint:disable-next-line no-var-requires */
const fetch = require('jest-fetch-mock');

jest.mock('../../../../../configs/env', () => ({
  GISIDA_MAPBOX_TOKEN: 'hunter2',
  GISIDA_TIMEOUT: 3000,
  HIDDEN_MAP_LEGEND_ITEMS: ['Partially Sprayed', 'Not Visited'],
  SUPERSET_IRS_REPORTING_INDICATOR_ROWS: 'zambia2019',
  SUPERSET_IRS_REPORTING_INDICATOR_STOPS: 'zambia2019',
  SUPERSET_IRS_REPORTING_JURISDICTIONS_DATA_SLICES: '11,12',
  SUPERSET_IRS_REPORTING_PLANS_SLICE: '13',
  SUPERSET_IRS_REPORTING_STRUCTURES_DATA_SLICE: '14',
  SUPERSET_JURISDICTIONS_SLICE: 1,
  SUPERSET_MAX_RECORDS: 2000,
}));

/** register the reducers */
reducerRegistry.register(IRSPlansReducerName, IRSPlansReducer);
reducerRegistry.register(jurisdictionReducerName, jurisdictionReducer);
reducerRegistry.register(GenericJurisdictionsReducerName, GenericJurisdictionsReducer);
reducerRegistry.register(genericStructuresReducerName, genericStructuresReducer);

/** set up data in the store needed for this view */
const focusAreaData = superset.processData(fixtures.ZambiaFocusAreasJSON) || [];
const structureData = superset.processData(fixtures.ZambiaStructuresJSON) || [];
const jurisdictionData = superset.processData(fixtures.ZambiaAkros1JSON) || [];
const jurisdiction2Data = superset.processData(fixtures.ZambiaKMZ421JSON) || [];

store.dispatch(fetchGenericJurisdictions('zm-focusAreas', focusAreaData));
store.dispatch(fetchGenericStructures('zm-structures', structureData));
store.dispatch(fetchJurisdictions(jurisdictionData));
store.dispatch(fetchJurisdictions(jurisdiction2Data));

const history = createBrowserHistory();

describe('components/IRS Reports/IRSReportingMap', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders map legend correctly with hidden items', async () => {
    fetch.mockResponseOnce(JSON.stringify({}));
    const mock: any = jest.fn();

    const supersetServiceMock: any = jest.fn();
    supersetServiceMock.mockImplementation(async () => []);

    const focusArea = getGenericJurisdictionByJurisdictionId(
      store.getState(),
      'zm-focusAreas',
      '0dc2d15b-be1d-45d3-93d8-043a3a916f30'
    );

    const structures = getGenericStructures(
      store.getState(),
      'zm-structures',
      '0dc2d15b-be1d-45d3-93d8-043a3a916f30'
    );

    const jurisdiction = getJurisdictionById(
      store.getState(),
      '0dc2d15b-be1d-45d3-93d8-043a3a916f30'
    );

    const props = {
      focusArea,
      history,
      jurisdiction,
      location: mock,
      match: {
        isExact: true,
        params: {
          jurisdictionId: '0dc2d15b-be1d-45d3-93d8-043a3a916f30',
          planId: (plans[0] as GenericPlan).plan_id,
        },
        path: `${REPORT_IRS_PLAN_URL}/:planId/:jurisdictionId/${MAP}`,
        url: `${REPORT_IRS_PLAN_URL}/${
          (plans[0] as GenericPlan).plan_id
        }/0dc2d15b-be1d-45d3-93d8-043a3a916f30/${MAP}`,
      },
      plan: plans[0] as GenericPlan,
      service: supersetServiceMock,
      structures,
    };
    const wrapper = mount(
      <Router history={history}>
        <IRSReportingMap {...props} />
      </Router>
    );
    await flushPromises();
    wrapper.update();
    expect(wrapper.find('.sidebar-legend-item').length).toEqual(3);
    expect(toJson(wrapper.find('.sidebar-legend-item'))).toMatchSnapshot('Legend items');
    wrapper.unmount();
  });
});
