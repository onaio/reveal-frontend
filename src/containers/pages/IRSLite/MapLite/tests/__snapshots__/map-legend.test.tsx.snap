// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/IRS Reports/IRSReportingMap renders map legend correctly with hidden items: Legend items 1`] = `
Array [
  <div
    className="sidebar-legend-item"
    key="0"
  >
    <span
      className="sidebar-legend-color"
      style={
        Object {
          "backgroundColor": "#68BB0C",
        }
      }
    />
    <span
      className="sidebar-legend-label"
    >
      Complete
    </span>
  </div>,
  <div
    className="sidebar-legend-item"
    key="1"
  >
    <span
      className="sidebar-legend-color"
      style={
        Object {
          "backgroundColor": "#F12525",
        }
      }
    />
    <span
      className="sidebar-legend-label"
    >
      Not Sprayed
    </span>
  </div>,
  <div
    className="sidebar-legend-item"
    key="4"
  >
    <span
      className="sidebar-legend-color"
      style={
        Object {
          "backgroundColor": "#000",
        }
      }
    />
    <span
      className="sidebar-legend-label"
    >
      Not Eligible
    </span>
  </div>,
]
`;
